linera-core/src/join_set_ext.rs

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! An extension trait to allow determining at compile time how tasks are spawned on the Tokio
+//! runtime.
+//!
+//! In most cases the [`Future`] task to be spawned should implement [`Send`], but that's
+//! not possible when compiling for `wasm32-unknown-unknown`. In that case, the task is
+//! spawned in a [`LocalSet`][`tokio::tast::LocalSet`].
+
+use std::future::Future;
+
+use tokio::{
+    sync::oneshot,
+    task::{AbortHandle, JoinSet},
+};
+
+/// An extension trait for the [`JoinSet`] type.
+#[cfg(not(target_arch = "wasm32"))]
+pub trait JoinSetExt: Sized {
+    /// Spawns a `future` task on this [`JoinSet`] using [`JoinSet::spawn`].
+    ///
+    /// Returns a [`oneshot::Receiver`] to receive the `future`'s output, and an
+    /// [`AbortHandle`] to cancel execution of the task.
+    fn spawn_task<F>(&mut self, future: F) -> (oneshot::Receiver<F::Output>, AbortHandle)
+    where
+        F: Future + Send + 'static,
+        F::Output: Send;
+
+    /// Awaits all tasks spawned in this [`JoinSet`].
+    fn await_all_tasks(&mut self) -> impl Future<Output = ()> + Send;
+
+    /// Reaps tasks that have finished.
+    fn reap_finished_tasks(&mut self);
+}
+
+/// An extension trait for the [`JoinSet`] type.
+#[cfg(target_arch = "wasm32")]
+pub trait JoinSetExt: Sized {
+    /// Spawns a `future` task on this [`JoinSet`] using [`JoinSet::spawn_local`].
+    ///
+    /// Returns a [`oneshot::Receiver`] to receive the `future`'s output, and an
+    /// [`AbortHandle`] to cancel execution of the task.
+    fn spawn_task<F>(&mut self, future: F) -> (oneshot::Receiver<F::Output>, AbortHandle)
+    where
+        F: Future + 'static;
+
+    /// Awaits all tasks spawned in this [`JoinSet`].
+    fn await_all_tasks(&mut self) -> impl Future<Output = ()>;
+
+    /// Reaps tasks that have finished.
+    fn reap_finished_tasks(&mut self);
+}
+
+#[cfg(not(target_arch = "wasm32"))]
```

Comment:
You could put that inside the `spawn_task` implementation, to deduplicate code and highlight the place where the implementations actually differ.

Reply:
They importantly differ on the `Send` bound — so this has to go outside the implementation.  But it could definitely be on the method rather than the trait, I think.

Reply:
(note: I also missed @Twey's comment here, thanks for clearing it up earlier :raised_hands:)